src/main/kotlin/org/rust/ide/presentation/RsPsiRendering.kt

```diff
@@ -0,0 +1,172 @@
+    return buildString {
+        append("(")
+        val selfParameter = list.selfParameter
+        val valueParameterList = list.valueParameterList
+        if (selfParameter != null) {
+            val typeReference = selfParameter.typeReference
+            if (typeReference != null) {
+                append("self: ")
+                appendTypeReference(typeReference)
+            } else {
+                if (selfParameter.isRef) {
+                    append("&")
+                    selfParameter.lifetime?.let {
+                        append(it.referenceName)
+                        append(" ")
+                    }
+                    append(if (selfParameter.mutability.isMut) "mut " else "")
+                }
+                append("self")
+            }
+            if (valueParameterList.isNotEmpty()) {
+                append(", ")
+            }
+        }
+        valueParameterList.joinToWithBuffer(this, separator = ", ") { sb ->
+            sb.append(patText ?: "_")
+            sb.append(": ")
+            typeReference?.let { sb.appendTypeReference(it) }
+        }
+        append(")")
+    }
+}
+
+private fun renderTypeReference(ref: RsTypeReference): String =
+    StringBuilder().apply { appendTypeReference(ref) }.toString()
+
+private fun StringBuilder.appendTypeReference(ref: RsTypeReference) {
+    val type = ref.typeElement
+    when (type) {
+        is RsTupleType -> type.typeReferenceList.joinToWithBuffer(this, ", ", "(", ")") { it.appendTypeReference(this) }
+
+        is RsBaseType -> {
+            val kind = type.kind
+            when (kind) {
+                BtUnit -> append("()")
+                BtNever -> append("!")
+                BtUnderscore -> append("_")
+                is BtPath -> appendPath(kind.path)
+            }
+        }
+
+        is RsRefLikeType -> {
+            if (type.isPointer) {
+                append(if (type.mutability.isMut) "*mut " else "*const ")
+            } else if (type.isRef) {
+                append("&")
+                type.lifetime?.let {
+                    append(it.referenceName)
+                    append(" ")
+                }
+                if (type.mutability.isMut) append("mut ")
+            }
+            appendTypeReference(type.typeReference)
+        }
+
+        is RsArrayType -> {
+            append("[")
+            appendTypeReference(type.typeReference)
+            if (!type.isSlice) {
+                append("; ")
+                append(type.arraySize) // may trigger resolve
+            }
+            append("]")
+        }
+
+        is RsFnPointerType -> {
+            append("fn")
+            appendValueParameterListTypes(type.valueParameterList.valueParameterList)
+            appendRetType(type.retType)
+        }
+
+        is RsTraitType -> {
+            append(if (type.isImpl) "impl " else "dyn ")
+            type.polyboundList.joinToWithBuffer(this, "+") { sb ->
```

Comment:
Looks like we need use `" + "` here.
![image](https://user-images.githubusercontent.com/2539310/48739987-c7efc900-ec66-11e8-81ec-338549e3d471.png)

Reply:
It's same as `impl Trait` [ty rendering](https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/ide/presentation/TypeRendering.kt#L112)

Reply:
So we find two places where we need to add spaces around `+` 😄 
Now it looks inconsistently with other separators like `:`, `,` and `->` where we use spaces